upcloud/managed_database.go

```diff
@@ -0,0 +1,693 @@
+type ManagedDatabaseProperties map[string]interface{}
+
+// Get returns a property value by name. The underlying map is initialised if it's nil
+func (m *ManagedDatabaseProperties) Get(name string) interface{} {
+	if m == nil {
+		return nil
+	}
+	if *m == nil {
+		*m = make(ManagedDatabaseProperties)
+	}
+	return (*m)[name]
+}
+
+// GetString returns a string property value. An empty string is returned for non-existing or not-a-string values.
+// The underlying map is initialised if it's nil.
+func (m *ManagedDatabaseProperties) GetString(name string) string {
+	v, _ := m.Get(name).(string)
+	return v
+}
+
+// GetInt returns an integer property value. A zero is returned for non-existing or not-an-integer values.
+// The underlying map is initialised if it's nil.
+func (m *ManagedDatabaseProperties) GetInt(name string) int {
+	v, _ := m.Get(name).(int)
+	return v
+}
+
+// GetBool returns a boolean property value. A false is returned for non-existing or not-a-boolean values.
+// The underlying map is initialised if it's nil.
+func (m *ManagedDatabaseProperties) GetBool(name string) bool {
+	v, _ := m.Get(name).(bool)
+	return v
+}
+
+// GetStringSlice returns a string-slice property value. A nil slice is returned for non-existing or not-a-string-slice values.
+// The underlying map is initialised if it's nil.
+func (m *ManagedDatabaseProperties) GetStringSlice(name string) []string {
+	if slice, ok := m.Get(name).([]string); ok {
+		return slice
+	}
+	slice, ok := m.Get(name).([]interface{})
+	if !ok {
+		return nil
+	}
+	stringSlice := make([]string, len(slice))
+	for i, iv := range slice {
+		v, ok := iv.(string)
+		if !ok {
+			return nil
+		}
+		stringSlice[i] = v
+	}
+	return stringSlice
+}
+
+// GetAutoUtilityIPFilter returns the state of automatic utility network IP filtering.
+// See upcloud.ManagedDatabasePropertyAutoUtilityIPFilter for more information.
+func (m *ManagedDatabaseProperties) GetAutoUtilityIPFilter() bool {
+	return m.GetBool(ManagedDatabasePropertyAutoUtilityIPFilter)
+}
+
+// GetIPFilter returns a slice of allowed hosts or networks.
+// See upcloud.ManagedDatabasePropertyIPFilter for more information.
+func (m *ManagedDatabaseProperties) GetIPFilter() []string {
+	return m.GetStringSlice(ManagedDatabasePropertyIPFilter)
+}
+
+// GetPublicAccess returns the state of public access to the service.
+// See upcloud.ManagedDatabasePropertyPublicAccess for more information.
+func (m *ManagedDatabaseProperties) GetPublicAccess() bool {
+	return m.GetBool(ManagedDatabasePropertyPublicAccess)
+}
+
+type ManagedDatabaseLogs struct {
+	// Offset describes the next available offset. Use this to query more logs.
+	Offset string `json:"offset"`
+	Logs   []ManagedDatabaseLogEntry
+}
+
+type ManagedDatabaseLogEntry struct {
+	Hostname string    `json:"hostname"`
+	Message  string    `json:"msg"`
+	Time     time.Time `json:"time"`
+	Unit     string    `json:"unit"`
```

Comment:
Unit could maybe use a comment? It's not clear to me (at least) what a unit refers to in the context of a log entry.

Reply:
I think I will open a discussion on changing this in the API to `service`. Unit is certainly very misleading.

Reply:
Ah, right, makes sense now :D

Reply:
implemented